diff --git a/fn/index.d.ts b/fn/index.d.ts
index 2e7a941b133656bd1c8625693f7dbfec56a65b40..700b51352e322821d4516a5d9dbcfc1d23102754 100755
--- a/fn/index.d.ts
+++ b/fn/index.d.ts
@@ -73,7 +73,7 @@ export {
 export { default as lerp } from '../src/interpolate/lerp';
 export { default as samples } from '../src/samples';
 export { default as displayable } from '../src/displayable';
-export { clampRgb, clampChroma } from '../src/clamp';
+export { clampRgb, clampChroma, toGamut, clampGamut } from '../src/clamp';
 export { default as nearest } from '../src/nearest';
 export { useMode, getMode, useParser, removeParser } from '../src/modes';
 export { default as parse } from '../src/parse';
diff --git a/index.d.ts b/index.d.ts
index a3bbc4fc81e4c87f0203fc094842f011b9671c72..cc61e4d4cb3c47926f717b0da28854d8a93edc12 100755
--- a/index.d.ts
+++ b/index.d.ts
@@ -80,7 +80,7 @@ export {
 export { default as lerp } from './src/interpolate/lerp';
 export { default as samples } from './src/samples';
 export { default as displayable } from './src/displayable';
-export { clampRgb, clampChroma } from './src/clamp';
+export { clampRgb, clampChroma, toGamut, clampGamut } from './src/clamp';
 export { default as nearest } from './src/nearest';
 export { useMode, getMode, useParser, removeParser } from './src/modes';
 export { default as parse } from './src/parse';
diff --git a/src/clamp.d.ts b/src/clamp.d.ts
index cf5bd58d016f0d89a71cf8dbcf42c4671a325820..0b17e28ba695b4683b9754107214d59c47fb2554 100755
--- a/src/clamp.d.ts
+++ b/src/clamp.d.ts
@@ -1,7 +1,16 @@
-import { Color, Mode } from './common';
+import { Color, Mode } from "./common";
+import { DiffFn } from "./difference";
 
 export function clampRgb(color: string): Color | undefined;
 export function clampRgb<C extends Color>(color: C): C;
 
 export function clampChroma(color: string, mode?: Mode): Color | undefined;
 export function clampChroma<C extends Color>(color: C, mode?: Mode): C;
+
+export function toGamut(
+  dest?: Mode,
+  mode?: Mode,
+  delta?: DiffFn,
+  jnd?: number
+): (color: string | Color) => Color;
+export function clampGamut(mode?: Mode): (color: string | Color) => Color;
